### PR TITLE
Introducing communication with alchemist-server through socket connection

### PR DESCRIPTION
**Pull request by slashmili**
By this change Alchemist.Server can listen to a port and reply the command through socket connection:

Starting the server
```
$ cd elixir_project
$ elixir path/to/alchemist-server/run.exs --env=dev --listen
ok|localhost:55580
```
Then the client connects and talk to the server like this:

```
$ nc localhost 56640
PING
PONG
END-OF-PING
COMP { "def", [ context: Elixir, imports: [Enum], aliases: [{MyList, List}] ] }
def
defexception/1
defoverridable/1
defstruct/1
def/2
defdelegate/2
defmacro/2
defmacrop/2
defmodule/2
defp/2
defprotocol/2
defimpl/3
END-OF-COMP
```
This changes allows us to introduce vim integration, if your interested to see how it would work, take look at [alchemist.vim](https://github.com/slashmili/alchemist.vim)

### CODE REVIEW

**Comment by slashmili** [comment]
@tonini I'm done! Can you review the code and give me your feedback?

File: lib/server/socket.exs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+Code.require_file "../helpers/process_commands.exs", __DIR__
+
+defmodule Alchemist.Server.Socket do
+  use Application
```
Comment (lexmag) [inline lib/server/socket.exs]: This looks very weird that it's an application.
As I see it doesn't need to be so.

Comment (slashmili) [inline lib/server/socket.exs]: When I wanted to mix gen_tcp and process in OTP I had two options:
1. handle it like `lib/server/io.exs`, which means run a GenServer and accept new connections in `handle_info(:timeout, state)` which I didn't find it nice.
2. While I was looking at docs to see how people implement this, I came across [Task and gen_tcp](http://elixir-lang.org/getting-started/mix-otp/task-and-gen-tcp.html) in elixir getting started documents.

Option 2 looks decent to me, it also handles corner cases. That's why I choose this approach.

If you have other suggestion, I'd be happy to dig in and implement. Then we can compare and choose the best option

Comment (slashmili) [inline lib/server/socket.exs]: You are correct! I removed `use Application`

File: lib/server/socket.exs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    |> ProcessCommands.process(env, io_string)
+
+    {:ok, {_, output}} = StringIO.close(io_string)
+    write_line(output, socket)
+
+    serve(socket, env)
+  end
+
+  defp read_line(socket) do
+    #TODO: handle {:error, :closed}
```
Comment (slashmili) [inline lib/server/socket.exs]: Anyone has suggestion for this part?

I wanted to handle this case and shutdown the Task but

```elixir
case :gen_tcp.recv(socket, 0) do
  {:ok, data} -> data
  {:error, :closed} -> Task.shutdown(self)
end
```

Didn't work as I expected

**Comment by slashmili** [comment]
FYI I wrote [a client](https://github.com/slashmili/alchemist.vim/blob/develop/alchemist_client) for the server and works like a charm!

**Comment by tonini** [comment]
Hi @slashmili 

First, thanks a lot for all your work you invested into this, it's really great and I like the idea. :)

I'm still no reviewed the whole thing but the first thing I saw and I'm not sure about it the `device` argument which is now needed everywhere. 

What do you think about using `ets` like elixir does it for storing the `iex` server? Have a look here https://github.com/elixir-lang/elixir/search?utf8=%E2%9C%93&q=autocomplete_server

/cc @lexmag @gausby @msaraiva It would be great if you guys also could review the PR too, you guys using the alchemist-server too and @msaraiva has great experience with using it and extend it as well.

Cheers

And now let me eat my chocolate bunny please ;-)

**Comment by slashmili** [comment]
Thanks for you feedback. I'm not quite sure about your suggestion. In network connection case, there might be multiple clients using the same server on the same time and I'm not sure if registering in a process in `ets` would help. Because there would be multiple process and each one answers to different client.

I can make change for the APIs so they return the result back to `Alchemist.Server.IO` and `Alchemist.Server.Socket`. Each of these servers know how to put the data back to their own clients. What do you think about this idea?

File: lib/api/docl.exs [hunk 3]
```diff
@@ -7,53 +8,55 @@ defmodule Alchemist.API.Docl do
...
-    Application.put_env(:iex, :colors, [enabled: true])
-
+  def request(args, device) do
     args
     |> normalize
-    |> process
+    |> process(device)
 
-    IO.puts "END-OF-DOCL"
+    IO.puts device, "END, func_puts-OF-DOCL"
```
Comment (tonini) [inline lib/api/docl.exs]: @slashmili I guess this one is wrong? Shouldn't it be `IO.puts device, "END-OF-DOCL"` instead?

Comment (slashmili) [inline lib/api/docl.exs]: My mistake! I'll change it.

File: lib/server/socket.exs (hunk 2 above)
Comment (tonini) [inline lib/server/socket.exs]: What's the status about that, did you figured it out?

File: lib/api/docl.exs (hunk 3 above)
Comment (slashmili) [inline lib/api/docl.exs]: Fixed and added a test for the case

File: lib/server/socket.exs (hunk 2 above)
Comment (slashmili) [inline lib/server/socket.exs]: No, I'll take a look at it this weekend.

Comment (tonini) [inline lib/server/socket.exs]: I guess if this one is done we could merge. :+1:

**Comment by tonini** [comment]
@slashmili It would be great if you could squash all commits to one with a proper commit message. https://github.com/tonini/alchemist.el/blob/master/CONTRIBUTING.md#commit-messages

**Comment by slashmili** [comment]
Sure will do. Sorry for polluting the history  :)

**Comment by tonini** [comment]
haha don't worry :) I just stumbled over it and thought it might be a good idea given that this PR brings new cool things.

**Comment by msaraiva** [comment]
@tonini I'll take a look at it.
@slashmili I just love this feature. As you can see, it was something already planned for atom-elixir as well: https://github.com/msaraiva/atom-elixir/issues/16

**Comment by slashmili** [comment]
@msaraiva great! We are almost there.

Just a bit background about communication to the server, I wanted to use unix socket but apparently [gen_tcp](http://erlang.org/doc/man/gen_tcp.html) doesn't support  AF_UNIX.

Instead `Alchemist.Server.Socket` is listening to first available port and print it out to `STDOUT` and the client can read the connection details from `STDOUT` and connect to the server.

**Comment by slashmili** [comment]
@tonini for the case that you mentioned the `device` needs to be passed in all the function, it's a valid point. I [made changes](https://github.com/slashmili/alchemist-server/commit/b72dd7d4296a0213c12e5ae618c436b8ae0bf177) in PING command, It covers your concern but it means that each commands should return the result as string instead of printing it out. If you are fine with that I'll refactor the rest of the commands to work like that.

As your suggestion for squashing the commits, I'll do it when we are close to the merge.

File: lib/server/socket.exs (hunk 2 above)
Comment (slashmili) [inline lib/server/socket.exs]: I the [handle case](https://github.com/slashmili/alchemist-server/commit/85af3cf01b3ebf7adcaf03510065cc4299d93fd9) by exiting from the loop if `:closed` is received

**Comment by tonini** [comment]
@slashmili I think that would be a good way if you're also ok with it.

**Comment by slashmili** [comment]
Ok, I'll make the change then.

**Comment by slashmili** [comment]
@tonini I [made the changes](https://github.com/tonini/alchemist-server/pull/8/commits/4bc8377e680235fdd9a1aa9cc54146032ecb4832)

Squashed the changes in a commit and created PR #9

**Pull request closed without merging**